doc/man/man1/qlimit.1
'\" t
.\"___INFO__MARK_BEGIN__
.\"
.\" Copyright: 2004 by Sun Microsystems, Inc.
.\"
.\"___INFO__MARK_END__
.\"
.\" $RCSfile: qlimit.1,v $ 
.\"
.\"
.\" Some handy macro definitions [from Tom Christensen's man(1) manual page].
.\"
.de SB		\" small and bold
.if !"\\$1"" \\s-2\\fB\&\\$1\\s0\\fR\\$2 \\$3 \\$4 \\$5
..
.\"
.de T		\" switch to typewriter font
.ft CW		\" probably want CW if you don't have TA font
..
.\"
.de TY		\" put $1 in typewriter font
.if t .T
.if n ``\c
\\$1\c
.if t .ft P
.if n \&''\c
\\$2
..
.\"
.de M		\" man page reference
\\fI\\$1\\fR\\|(\\$2)\\$3
..
.TH QLIMIT 1 "$Date: 2006/10/24 08:05:01 $" "xxRELxx" "xxQS_NAMExx User Commands"
.SH NAME
qlimit \- show current usage of xxQS_NAMExx limitation rule sets
.\"
.\"
.SH SYNTAX
.B qlimit
[
.B -h wc_host|wc_hostgroup,...
] [
.B -help
] [
.B -l resource_name,...
] [
.B -u wc_user,...
] [
.B -P wc_project,...
] [
.B -pe wc_pe_name,...
] [
.B -q wc_cqueue,...
] [
.B -xml
]
.\"
.SH DESCRIPTION
.I qlimit
shows the current xxQS_NAMExx limitation rule sets. Only limitation rule sets
with a positive usage count or a static limit are printed.
.PP
Selection options allow you to filter for specific hosts, cluster queues,
projects, parallel environments (pe), resources or users.
Without any option
.I qlimit
will display a list of all limitation rule sets for the calling user.
.PP
.\"
.SH OPTIONS
.\"
.IP "\fB\-h wc_host|wc_hostgroup,...\fP"
.I 
Display only limitation rule sets that matches with the hosts
in the given wildcard host or hostgroup list. Find the definition
of \fBwc_host\fP and \fBwc_hostgroup\fP in
.M sge_types 1 .
.\"
.IP "\fB\-help\fP"
Prints a listing of all options.
.\"
.IP "\fB\-l resource_name,...\fP"
Display only limitation rule sets being matched with the resources in the
given resource list.
.\"
.IP "\fB\-u wc_user,...\fP"
Display only limitation rule sets being matched with the users in the
given wildcard user list. Find the definition
of \fBwc_user\fP in
.M sge_types 1 .
.\"
.IP "\fB\-P wc_project,...\fP"
Display only limitation rule sets being matched with the projects in the
given wildcard project list. Find the definition
of \fBwc_project\fP in
.M sge_types 1 .
.\"
.IP "\fB\-pe wc_pe_name,...\fP"
Display only limitation rule sets being matched with the parallel environments (pe)
in the given wildcard pe list. Find the definition
of \fBwc_pe_name\fP in
.M sge_types 1 .
.\"
.IP "\fB\-q wc_cqueue,...\fP"
Display only limitation rule sets being matched with the queues in the
given wildcard cluster queue list. Find the definition
of \fBwc_cqueue\fP in
.M sge_types 1 .
.\"
.IP "\fB\-xml \fP"
This option can be used with all other options and changes the output to XML. The
schema used is referenced in the XML output. The output is printed to stdout.
.\"
.\"
.SH "OUTPUT FORMATS"
A line is printed for every limitation rule with a positive usage count or a static resource.
The line consists of
.IP "\(bu" 3n
the limitation rule - rule set name/rule name or number of rule in ruleset
.IP "\(bu" 3n
the limit - resource name, the number of used and available job slots
.IP "\(bu" 3n
the effective limitation rule set filter
.\"
.\"
.SH "ENVIRONMENTAL VARIABLES"
.\" 
.IP "\fBxxQS_NAME_Sxx_ROOT\fP" 1.5i
Specifies the location of the xxQS_NAMExx standard configuration
files.
.\"
.IP "\fBxxQS_NAME_Sxx_CELL\fP" 1.5i
If set, specifies the default xxQS_NAMExx cell. To address a xxQS_NAMExx
cell
.I qlimit
uses (in the order of precedence):
.sp 1
.RS
.RS
The name of the cell specified in the environment 
variable xxQS_NAME_Sxx_CELL, if it is set.
.sp 1
The name of the default cell, i.e. \fBdefault\fP.
.sp 1
.RE
.RE
.\"
.IP "\fBxxQS_NAME_Sxx_DEBUG_LEVEL\fP" 1.5i
If set, specifies that debug information
should be written to stderr. In addition the level of
detail in which debug information is generated is defined.
.\"
.IP "\fBxxQS_NAME_Sxx_QMASTER_PORT\fP" 1.5i
If set, specifies the tcp port on which
.M xxqs_name_sxx_qmaster 8
is expected to listen for communication requests.
Most installations will use a services map entry for the
service "sge_qmaster" instead to define that port.
.\"
.\"
.SH FILES
.nf
.ta \w'<xxqs_name_sxx_root>/     'u
\fI<xxqs_name_sxx_root>/<cell>/common/act_qmaster\fP
	xxQS_NAMExx master host file
.ta \w'<xxqs_name_sxx_root>/     'u
\fI<xxqs_name_sxx_root>/<cell>/common/xxqs_name_sxx_qlimit\fP
	cluster qlimit default options
\fI$HOME/.xxqs_name_sxx_qlimit\fP	
	user qlimit default options
.fi
.\"
.\"
.SH "SEE ALSO"
.M xxqs_name_sxx_intro 1 ,
.M qalter 1 ,
.M qconf 1 ,
.M qhold 1 ,
.M qmod 1 ,
.M qstat 1 ,
.M qsub 1 ,
.M queue_conf 5 ,
.M xxqs_name_sxx_execd 8 ,
.M xxqs_name_sxx_qmaster 8 ,
.M xxqs_name_sxx_shepherd 8 .
.\"
.\"
.SH "COPYRIGHT"
See
.M xxqs_name_sxx_intro 1
for a full statement of rights and permissions.


Changelog
AS-2006-10-30-1: Bugfix:    dist/util/sge_request referred to cod_request(5)
                            man page instead of sge_request(5)
                 Bugster:   6325359


AA-2006-10-30-0: Bugfix:    o repaired broken build on certain archs through sge_lock fix
                            o fixed error messages for PEM_BYTE code
                 Review:    RD

RH-2006-10-30-0: Bugfix:    o jgdi: QueueInstanceSummary did not work with a queue filter
                              (JB_hard_resource_list has been reduced)
                 Review:    AA

AA-2006-10-27-0: Bugfix:    o changed sge_lock to use static initializers
                              and removed unused functions sge_setup_lock_service() etc.
                            o fix for subscribe/unsubscribe after being registered
                 Review:    RD

AH-2006-10-26-2: Bugfix:    Fix an issue with cl_com_set_synchron_receive_timeout()
                            be called with a NULL commlib handle upon changes with
                            sched_conf(5)
                 Review:    AA

AH-2006-10-26-1: Enhancem.  Added commlib wrapper functions to allow qmaster 
                            message buffer be monitored through dtrace
                 Review:    CR

AH-2006-10-26-0: Bugfix:    qmaster loggings "scheduler tried to remove a 
                            incomplete" and "scheduler send a order for a 
                            changed user/project"
                 Bugster:   6486125
                 Issue:     2108
                 Review:    JG

AA-2006-10-25-0: Enhancem.  o openss-0.9.7l
                            o aimk fix for darwin
                            o execd USE_QSUB_GID problem fix
                            o drmaa build junit.jar dependency
                            o darwin JNI_CreateJavaVM_Impl due to
                              missing JNI_CreateJavaVM
                            o jni.h wrong dependency
                 Review:    JG

RD-2006-10-24-0: Enhancem.  o added lirs qconf options to qconf(1)
                            o added qlimit(1) manpage
                            o added sge_limitation_rule_set(5) man page
                            o unified format of all man pages
                 Review:    DT

AH-2006-10-23-1: Bugfix:    Immediate array jobs that exceed the number of 
                            available resources are not rejected completely
                 Bugtraq:   6480138
                 Review:    AA

AH-2006-10-23-0: Bugfix:    Discrepancy between load values reported by 
                            Gridengine and from the HP-UX 64 bit env
                 Bugtraq:   6470048
                 Review:    JG

RH-2006-10-23-0: Bugfix:    jgdi: Cat and paste error prevents handling of
                            ExecHostAddEvent
                 Review:    AA
 
GH-2006-10-20-0: Bugfix:    Account string does not accept the "|" character
                 Bugtraq:   6475282
                 Review:    AH

RD-2006-10-19-0: Bugfix:    - Execd ignores in certain cases JOB_EXECUTION request
                            - removed some redundant code
                 Bugster:   6483941
                 Issue:     2106
                 Review:    JG

AH-2006-10-19-0: Bugfix:    Ensure events emitted upon qmod -d/-e get always
                            sent as a whole to the scheduler.
                 Review:    RD

JG-2006-10-19-0: Bugfix:    Fixed broken build on irix and darwin.
                 Review:    RD

AH-2006-10-16-3: Enhancem.: Added RUNPATH link flag to SHARED_LFLAGS to fix problem
                            of libspoolb.so depending on LD_LIBRARY_PATH for 
                            libdb-4.4.so.
                 Bugster:   6456290
                 Issue:     2090
                 
AH-2006-10-16-1: Enhancem.: Fixed NO_SGE_COMPILE_DEBUG setting to cause rmon-macros 
                            be inactive in pack.c, cull_list.c and cull_hash.c

AH-2006-10-16-0: Enhancem.: Forgot inst_common.sh with AH-2006-10-12-0
                 Bugster:   6456290
                 Issue:     2090
                 Review:    JG

AH-2006-10-12-0: Enhancem.: Added distinst -shlibpath to allow enforcement of 
                            LD_LIBRARY_PATH setting even with Solaris/Linux where 
                            actually $ORIGIN relative RUNPATH is used instead
                 Bugster:   6456290
                 Issue:     2090
                 Review:    AS

AH-2006-10-10-0: Enhancem:  Added #pragma no_inline() to prevent inlining of some 
                            functions as required for dtrace(1) pid provider. This 
                            change is explained in dtrace-doc CR #6480235.
                 Review:    RD

RD-2006-10-09-0: Bugfix:    Fixed build errors
                            - with non gnu make
                            - on darwin_x86
                            - make rules for windows
                 Review:    AA

JG-2006-10-09-0: Bugfix:    Fixed msg_juti.h.
                            aimk didn't fail on windows in case of errors while
                            building windows applications.

source/clients/qlimit/qlimit.c
/*___INFO__MARK_BEGIN__*/
/*************************************************************************
 * 
 *  The Contents of this file are made available subject to the terms of
 *  the Sun Industry Standards Source License Version 1.2
 * 
 *  Sun Microsystems Inc., March, 2001
 * 
 * 
 *  Sun Industry Standards Source License Version 1.2
 *  =================================================
 *  The contents of this file are subject to the Sun Industry Standards
 *  Source License Version 1.2 (the "License"); You may not use this file
 *  except in compliance with the License. You may obtain a copy of the
 *  License at http://gridengine.sunsource.net/Gridengine_SISSL_license.html
 * 
 *  Software provided under this License is provided on an "AS IS" basis,
 *  WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING,
 *  WITHOUT LIMITATION, WARRANTIES THAT THE SOFTWARE IS FREE OF DEFECTS,
 *  MERCHANTABLE, FIT FOR A PARTICULAR PURPOSE, OR NON-INFRINGING.
 *  See the License for the specific provisions governing your rights and
 *  obligations concerning the Software.
 * 
 *   The Initial Developer of the Original Code is: Sun Microsystems, Inc.
 * 
 *   Copyright: 2001 by Sun Microsystems, Inc.
 * 
 *   All Rights Reserved.
 * 
 ************************************************************************/
/*___INFO__MARK_END__*/
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <limits.h>
#include <math.h>
#include <float.h>
#include <pwd.h>

#include "basis_types.h"
#include "sge.h"

#include "sge_bootstrap.h"

#include "sge_gdi.h"
#include "sge_all_listsL.h"
#include "commlib.h"
#include "cull_xml.h"
#include "sig_handlers.h"
#include "sge_prog.h"
#include "sgermon.h"
#include "sge_feature.h"
#include "sge_unistd.h"
#include "sge_stdlib.h"
#include "cull_parse_util.h"
#include "parse.h"
#include "sge_host.h"
#include "sge_complex_schedd.h"
#include "sge_parse_num_par.h"
#include "sge_select_queue.h"
#include "qstat_printing.h"
#include "sge_range.h"
#include "load_correction.h"
#include "sge_conf.h"
#include "msg_common.h"
#include "msg_clients_common.h"
#include "msg_qlimit.h"
#include "sge_string.h"
#include "sge_hostname.h"
#include "sge_log.h"
#include "sge_answer.h"
#include "sge_qinstance.h"
#include "sge_qinstance_state.h"
#include "sge_qinstance_type.h"
#include "sge_ulong.h"
#include "sge_centry.h"
#include "sgeobj/sge_schedd_conf.h"
#include "sge_mt_init.h"
#include "sge_qlimit.h"
#include "sge_object.h"
#include "uti/sge_profiling.h"
#include "uti/sge_uidgid.h"
#include "uti/setup_path.h"
#include "read_defaults.h"
#include "uti/sge_io.h"

#ifdef TEST_GDI2
#include "sge_gdi_ctx.h"
#endif

static report_handler_t* create_xml_report_handler(lList **alpp);

static int xml_report_finished(report_handler_t* handler, lList **alpp);
static int xml_report_started(report_handler_t* handler, lList **alpp);

static int xml_report_limit_rule_begin(report_handler_t* handler, const char* limit_name, lList **alpp);
static int xml_report_limit_rule_finished(report_handler_t* handler, const char *limit_name, lList **alpp);
static int xml_report_limit_string_value(report_handler_t* handler, const char* name, const char *value, bool exclude, lList **alpp);
static int xml_report_resource_value(report_handler_t* handler, const char* resource, const char* limit, const char*value, lList **alpp);
static int destroy_xml_report_handler(report_handler_t** handler, lList **alpp);

static int xml_report_started(report_handler_t* handler, lList **alpp) {
   printf("<?xml version='1.0'?>\n");
   printf("<qlimit_result xmlns:xsd=\"http://www.w3.org/2001/XMLSchema\">\n");
   return QLIMIT_SUCCESS;
}

static int xml_report_finished(report_handler_t* handler, lList **alpp) {
   printf("</qlimit_result>\n");
   return QLIMIT_SUCCESS;
}

static report_handler_t* create_xml_report_handler(lList **alpp) {
   
   report_handler_t* ret = (report_handler_t*)sge_malloc(sizeof(report_handler_t));
   if (ret == NULL ) {
      answer_list_add(alpp, "malloc of report_handler_t failed",
                            STATUS_EMALLOC, ANSWER_QUALITY_ERROR);
      return NULL;
   }
   ret->ctx = sge_malloc(sizeof(dstring));
   if (ret->ctx == NULL ) {
      answer_list_add(alpp, "malloc of dstring buffer failed",
                            STATUS_EMALLOC, ANSWER_QUALITY_ERROR);
      return NULL;
   }
   memset(ret->ctx, 0, sizeof(dstring));
   
   ret->report_started = xml_report_started;
   ret->report_finished = xml_report_finished;
   ret->report_limit_rule_begin = xml_report_limit_rule_begin;
   ret->report_limit_string_value = xml_report_limit_string_value;
   ret->report_limit_rule_finished = xml_report_limit_rule_finished;
   ret->report_resource_value = xml_report_resource_value;

   ret->destroy = destroy_xml_report_handler;
   return ret;
}

static int destroy_xml_report_handler(report_handler_t** handler, lList **alpp) {
   if (*handler != NULL ) {
      sge_dstring_free((dstring*)(*handler)->ctx);
      FREE((*handler)->ctx);
      FREE(*handler);
      *handler = NULL;
   }
   return QLIMIT_SUCCESS;
}

static int xml_report_limit_rule_begin(report_handler_t* handler, const char* limit_name, lList **alpp) {
   
  escape_string(limit_name, (dstring*)handler->ctx);
  printf(" <qlimit_rule name='%s'>\n", sge_dstring_get_string((dstring*)handler->ctx));
  sge_dstring_clear((dstring*)handler->ctx);
  return QLIMIT_SUCCESS;
}

static int xml_report_limit_string_value(report_handler_t* handler, const char *name, const char *value, bool exclude, lList **alpp) {
   escape_string(name, (dstring*)handler->ctx);
   if (exclude) {
      printf("   <x%s>", sge_dstring_get_string((dstring*)handler->ctx) );
   } else {
      printf("   <%s>", sge_dstring_get_string((dstring*)handler->ctx) );
   }

   sge_dstring_clear((dstring*)handler->ctx);
   escape_string(value, (dstring*)handler->ctx);
   printf("%s", sge_dstring_get_string((dstring*)handler->ctx));

   sge_dstring_clear((dstring*)handler->ctx);
   escape_string(name, (dstring*)handler->ctx);
   if (exclude) {
      printf("</x%s>\n", sge_dstring_get_string((dstring*)handler->ctx));
   } else {
      printf("</%s>\n", sge_dstring_get_string((dstring*)handler->ctx));
   }

   sge_dstring_clear((dstring*)handler->ctx);

   return QLIMIT_SUCCESS;
}

static int xml_report_limit_rule_finished(report_handler_t* handler, const char *limit_name, lList **alpp) {
  printf(" </qlimit_rule>\n");   
   return QLIMIT_SUCCESS;
}

static int xml_report_resource_value(report_handler_t* handler, const char* resource, const char* limit, const char *value, lList **alpp) {
   escape_string(resource, (dstring*)handler->ctx);
   printf("   <limit resource='%s' ", sge_dstring_get_string((dstring*)handler->ctx));
 
   sge_dstring_clear((dstring*)handler->ctx);
   escape_string(limit, (dstring*)handler->ctx);   
   printf("limit='%s'", sge_dstring_get_string((dstring*)handler->ctx));

   if (value != NULL) {
      sge_dstring_clear((dstring*)handler->ctx);
      escape_string(value, (dstring*)handler->ctx);   
      printf(" value='%s'", sge_dstring_get_string((dstring*)handler->ctx));
   }
   printf("/>\n");
   
   sge_dstring_clear((dstring*)handler->ctx);
   return QLIMIT_SUCCESS;
}

static bool sge_parse_from_file_qlimit(const char *file, lList **ppcmdline, lList **alpp);
static bool sge_parse_cmdline_qlimit(char **argv, lList **ppcmdline, lList **alpp);
static bool sge_parse_qlimit(lList **ppcmdline, lList **host_list, lList **resource_match_list,
                             lList **user_list, lList **pe_list, lList **project_list, lList **cqueue_list,
                             report_handler_t **report_handler, lList **alpp);
static bool qlimit_usage(FILE *fp);

extern char **environ;
                                      
/************************************************************************/
int main(int argc, char **argv)
{
   lList *pcmdline = NULL;
   lList *host_list = NULL;
   lList *resource_match_list = NULL;
   lList *user_list = NULL;
   lList *pe_list = NULL;
   lList *project_list = NULL;
   lList *cqueue_list = NULL;

   lList *alp = NULL;
   report_handler_t *report_handler = NULL;
   
   int qlimit_result = 0;

#ifdef TEST_GDI2   
   sge_gdi_ctx_class_t *ctx = NULL;
#endif

   DENTER_MAIN(TOP_LAYER, "qlimit");

   log_state_set_log_gui(true);

#ifdef TEST_GDI2
   if (sge_gdi2_setup(&ctx, QLIMIT, &alp) != AE_OK) {
      answer_list_output(&alp);
      sge_prof_cleanup();
      SGE_EXIT((void**)&ctx, 1);
   }
#else
   sge_mt_init();

   sge_gdi_param(SET_MEWHO, QLIMIT, NULL);
   if (sge_gdi_setup(prognames[QLIMIT], &alp) != AE_OK) {
      answer_list_output(&alp);
      sge_prof_cleanup();
      SGE_EXIT(NULL, 1);
   }
#endif

   sge_setup_sig_handlers(QLIMIT);
   
   /*
   ** stage 1 of commandline parsing
   */
   {
      dstring file = DSTRING_INIT;
#ifdef TEST_GDI2
      const char *user = ctx->get_username(ctx);
      const char *cell_root = ctx->get_cell_root(ctx);
#else
      const char *user = uti_state_get_user_name();
      const char *cell_root = path_state_get_cell_root();
#endif

      /* arguments from SGE_ROOT/common/sge_qlimit file */
      get_root_file_path(&file, cell_root, SGE_COMMON_DEF_QLIMIT_FILE);
      if (sge_parse_from_file_qlimit(sge_dstring_get_string(&file), &pcmdline, &alp) == true) {
         /* arguments from $HOME/.qlimit file */
         if (get_user_home_file_path(&file, SGE_HOME_DEF_QLIMIT_FILE, user, &alp)) {
            sge_parse_from_file_qlimit(sge_dstring_get_string(&file), &pcmdline, &alp);
         }
      }
      sge_dstring_free(&file); 

      if (alp) {
         answer_list_output(&alp);
         lFreeList(&pcmdline);
         sge_prof_cleanup();
         SGE_EXIT(NULL, 1);
      }
   }
   if (sge_parse_cmdline_qlimit(argv, &pcmdline, &alp) == false) {
      answer_list_output(&alp);
      lFreeList(&pcmdline);
      sge_prof_cleanup();
      SGE_EXIT(NULL, 1);
   }

   /*
   ** stage 2 of commandline parsing 
   */
   if (sge_parse_qlimit(&pcmdline, 
            &host_list,             /* -h host_list                  */
            &resource_match_list,   /* -l resource_request           */
            &user_list,             /* -u user_list                  */
            &pe_list,               /* -pe pe_list                   */
            &project_list,          /* -P project_list               */
            &cqueue_list,           /* -q wc_queue_list              */
            &report_handler,
            &alp) == false) {
      /*
      ** low level parsing error! show answer list
      */
      answer_list_output(&alp);
      lFreeList(&pcmdline);
      sge_prof_cleanup();
      SGE_EXIT(NULL, 1);
   }

#ifdef TEST_GDI2
   qlimit_result = qlimit_output(ctx, host_list, resource_match_list, user_list, pe_list, project_list, cqueue_list, &alp, report_handler);
#else
   qlimit_result = qlimit_output(NULL, host_list, resource_match_list, user_list, pe_list, project_list, cqueue_list, &alp, report_handler);
#endif
   
   if (report_handler != NULL ) {
      report_handler->destroy(&report_handler, &alp);
   }
   
   if ( qlimit_result != 0 ) {
      answer_list_output(&alp);
      sge_prof_cleanup();
      SGE_EXIT(NULL, 1);
   }
   sge_prof_cleanup();
   SGE_EXIT(NULL, 0); /* 0 means ok - others are errors */
   DEXIT;
   return 0;
}

/****** qlimit/qlimit_usage() **************************************************
*  NAME
*     qlimit_usage() -- displays qlimit help output
*
*  SYNOPSIS
*     static bool qlimit_usage(FILE *fp) 
*
*  FUNCTION
*     displays qlimit_usage for qlist client
*     note that the other clients use a common function
*     for this. output was adapted to a similar look.
*
*  INPUTS
*     FILE *fp - output file pointer
*
*  RESULT
*     static bool - true on success
*                   false on error
*
*  NOTES
*     MT-NOTE: qlimit_usage() is MT safe 
*
*******************************************************************************/
static bool 
qlimit_usage(FILE *fp)
{
   dstring ds;
   char buffer[256];

   DENTER(TOP_LAYER, "qlimit_usage");

   if (fp == NULL) {
      DRETURN(false);
   }

   sge_dstring_init(&ds, buffer, sizeof(buffer));

   fprintf(fp, "%s\n", feature_get_product_name(FS_SHORT_VERSION, &ds));
   fprintf(fp,"%s qlimit [options]\n", MSG_SRC_USAGE);
   fprintf(fp, "  [-help]                    %s\n", MSG_COMMON_help_OPT_USAGE);
   fprintf(fp, "  [-h host_list]             %s\n", MSG_QLIMIT_h_OPT_USAGE);
   fprintf(fp, "  [-l resource_attributes]   %s\n", MSG_QLIMIT_l_OPT_USAGE);
   fprintf(fp, "  [-u user_list]             %s\n", MSG_QLIMIT_u_OPT_USAGE);
   fprintf(fp, "  [-pe pe_list]              %s\n", MSG_QLIMIT_pe_OPT_USAGE);
   fprintf(fp, "  [-P project_list]          %s\n", MSG_QLIMIT_P_OPT_USAGE); 
   fprintf(fp, "  [-q wc_queue_list]         %s\n", MSG_QLIMIT_q_OPT_USAGE); 
   fprintf(fp, "  [-xml]                     %s\n", MSG_COMMON_xml_OPT_USAGE);

   DRETURN(true);
}

/****** qlimit/sge_parse_from_file_qlimit() ************************************
*  NAME
*     sge_parse_from_file_qlimit() -- parse qlimit command line options from
*                                     file
*
*  SYNOPSIS
*     static bool sge_parse_from_file_qlimit(const char *file, lList 
*     **ppcmdline, lList **alpp) 
*
*  FUNCTION
*     parses the qlimit command line options from file
*
*  INPUTS
*     const char *file  - file name
*     lList **ppcmdline - found command line options
*     lList **alpp      - answer list pointer
*
*  RESULT
*     static bool - true on success
*                   false on error
*
*  NOTES
*     MT-NOTE: sge_parse_from_file_qlimit() is MT safe 
*
*******************************************************************************/
static bool
sge_parse_from_file_qlimit(const char *file, lList **ppcmdline, lList **alpp)
{
   bool ret = true;

   DENTER(TOP_LAYER, "sge_parse_from_file_qlimit");
   if (ppcmdline == NULL) {
      ret = false;
   } else {
      if (!sge_is_file(file)) {
         /*
          * This is no error
          */
         DPRINTF(("file "SFQ" does not exist\n", file));
         ret = true;
      } else {
         char *file_as_string = NULL;
         int file_as_string_length;

         file_as_string = sge_file2string(file, &file_as_string_length);
         if (file_as_string == NULL) {
            answer_list_add_sprintf(alpp, STATUS_EUNKNOWN, 
                                    ANSWER_QUALITY_ERROR,
                                    MSG_ANSWER_ERRORREADINGFROMFILEX_S, file);
            ret = false;
         } else {
            char **token = NULL;

            token = stra_from_str(file_as_string, " \n\t");
            ret = sge_parse_cmdline_qlimit(token, ppcmdline, alpp);
         }
      }
   }  
   DRETURN(ret); 
}

/****** qlimit/sge_parse_cmdline_qlimit() **************************************
*  NAME
*     sge_parse_cmdline_qlimit() -- ??? 
*
*  SYNOPSIS
*     static bool sge_parse_cmdline_qlimit(char **argv, lList **ppcmdline, 
*     lList **alpp) 
*
*  FUNCTION
*     'stage 1' parsing of qlimit-options. Parses options
*     with their arguments and stores them in ppcmdline.
*
*  INPUTS
*     char **argv       - argument list
*     lList **ppcmdline - found arguments
*     lList **alpp      - answer list pointer
*
*  RESULT
*     static bool - true on success
*                   false on error
*
*  NOTES
*     MT-NOTE: sge_parse_cmdline_qlimit() is MT safe 
*
*******************************************************************************/
static bool sge_parse_cmdline_qlimit(char **argv, lList **ppcmdline, lList **alpp)
{
   char **sp;
   char **rp;
   DENTER(TOP_LAYER, "sge_parse_cmdline_qlimit");

   if (argv == NULL) {
      answer_list_add_sprintf(alpp, STATUS_ESEMANTIC, ANSWER_QUALITY_ERROR, MSG_NULLPOINTER);
      DRETURN(false);
   }

   rp = ++argv;
   while(*(sp=rp)) {
      /* -help */
      if ((rp = parse_noopt(sp, "-help", NULL, ppcmdline, alpp)) != sp)
         continue;
 
      /* -h option */
      if ((rp = parse_until_next_opt2(sp, "-h", NULL, ppcmdline, alpp)) != sp)
         continue;

      /* -l option */
      if ((rp = parse_until_next_opt2(sp, "-l", NULL, ppcmdline, alpp)) != sp)
         continue;

      /* -u option */
      if ((rp = parse_until_next_opt2(sp, "-u", NULL, ppcmdline, alpp)) != sp)
         continue;

      /* -pe option */
      if ((rp = parse_until_next_opt2(sp, "-pe", NULL, ppcmdline, alpp)) != sp)
         continue;

      /* -P option */
      if ((rp = parse_until_next_opt2(sp, "-P", NULL, ppcmdline, alpp)) != sp)
         continue;

      /* -q */
      if ((rp = parse_until_next_opt2(sp, "-q", NULL, ppcmdline, alpp)) != sp)
         continue;

      /* -xml */
      if ((rp = parse_noopt(sp, "-xml", NULL, ppcmdline, alpp)) != sp)
         continue;
      
      /* oops */
      qlimit_usage(stderr);
      answer_list_add_sprintf(alpp, STATUS_ESEMANTIC, ANSWER_QUALITY_ERROR, MSG_PARSE_INVALIDOPTIONARGUMENTX_S, *sp);
      DRETURN(false);
   }
   DRETURN(true);
}

/****** qlimit/sge_parse_qlimit() **********************************************
*  NAME
*     sge_parse_qlimit() -- parse qlimit options
*
*  SYNOPSIS
*     static bool sge_parse_qlimit(lList **ppcmdline, lList **host_list, lList 
*     **resource_list, lList **user_list, lList **pe_list, lList 
*     **project_list, lList **cqueue_list, report_handler_t **report_handler, 
*     lList **alpp) 
*
*  FUNCTION
*     'stage 2' parsing of qlimit-options. Gets the options from pcmdline
*
*  INPUTS
*     lList **ppcmdline                 - found command line options (from stage 1)
*     lList **host_list                 - parsed host list (-h option)
*     lList **resource_list             - parsed resource list (-l option)
*     lList **user_list                 - parsed user list (-u option)
*     lList **pe_list                   - parsed pe list (-pe option)
*     lList **project_list              - parsed project list (-P option)
*     lList **cqueue_list               - parsed queue list (-q option)
*     report_handler_t **report_handler - report handler for xml output
*     lList **alpp                      - answer list
*
*  RESULT
*     static bool - true on success
*                   false on error
*
*  NOTES
*     MT-NOTE: sge_parse_qlimit() is MT safe 
*
*******************************************************************************/
static bool
sge_parse_qlimit(lList **ppcmdline, lList **host_list, lList **resource_list,
                 lList **user_list, lList **pe_list, lList **project_list,
                 lList **cqueue_list, report_handler_t **report_handler, lList **alpp)
{
   bool showedhelp = false;
   u_long32 helpflag = 0;
   char *argstr = NULL;
   bool ret = true;
 
   DENTER(TOP_LAYER, "sge_parse_qlimit");
 
   /* Loop over all options. Only valid options can be in the
      ppcmdline list. 
   */
   while (lGetNumberOfElem(*ppcmdline)) {
      if (parse_flag(ppcmdline, "-help",  alpp, &helpflag)) {
         showedhelp = true;
         qlimit_usage(stdout);
         ret = false;
         break;
      }

      if (parse_multi_stringlist(ppcmdline, "-h", alpp, host_list, ST_Type, ST_name)) {
         /* 
         ** resolve hostnames and replace them in list
         */
         lListElem *ep = NULL;
         for_each(ep, *host_list) {
            sge_resolve_host(ep, ST_name);
         }
         continue;
      }

      if (parse_string(ppcmdline, "-l", alpp, &argstr)) {
         *resource_list = centry_list_parse_from_string(*resource_list, argstr, false);
         FREE(argstr);
         continue;
      }
      if (parse_multi_stringlist(ppcmdline, "-u", alpp, user_list, ST_Type, ST_name)) {
         continue;
      }
      if (parse_multi_stringlist(ppcmdline, "-pe", alpp, pe_list, ST_Type, ST_name)) {
         continue;
      }
      if (parse_multi_stringlist(ppcmdline, "-P", alpp, project_list, ST_Type, ST_name)) {
         continue;
      }
      if (parse_multi_stringlist(ppcmdline, "-q", alpp, cqueue_list, ST_Type, ST_name)) {
         continue;
      }
      if (parse_flag(ppcmdline, "-xml", alpp, &helpflag)) {
         *report_handler = create_xml_report_handler(alpp);
         continue;
      }
   }

   if (lGetNumberOfElem(*ppcmdline)) {
     if (showedhelp == false) {
       qlimit_usage(stderr);
     }
     answer_list_add_sprintf(alpp, STATUS_ESEMANTIC, ANSWER_QUALITY_ERROR, MSG_PARSE_TOOMANYOPTIONS);
     ret = false;
   }

   DRETURN(ret);
}
